src/vs/workbench/contrib/chat/browser/chatSessions.contribution.ts

```diff
@@ -794,58 +804,62 @@ export class ChatSessionsService extends Disposable implements IChatSessionsServ
 		};
 	}
 
-	public registerModelProgressListener(model: IChatModel, callback: () => void): void {
-		// Prevent duplicate registrations for the same model
-		if (this._registeredModels.has(model)) {
-			return;
-		}
-		this._registeredModels.add(model);
-
-		// Helper function to register listeners for a request
-		const registerRequestListeners = (request: IChatRequestModel) => {
-			if (!request.response || this._registeredRequestIds.has(request.id)) {
-				return;
+	public registerModelProgressListener(models: Iterable<IChatModel>): void {
+		for (const model of models) {
+			// Prevent duplicate registrations for the same model
+			if (this._registeredModels.has(model)) {
+				continue;
 			}
+			this._registeredModels.add(model);
+
+			// Get the session type from the model's sessionResource scheme
+			// Helper function to register listeners for a request
+			const registerRequestListeners = (request: IChatRequestModel) => {
+				const chatSessionType = model.sessionResource.scheme === Schemas.vscodeLocalChatSession ? 'local' : model.sessionResource.scheme;
+				if (!request.response || this._registeredRequestIds.has(request.id)) {
+					return;
+				}
 
-			this._registeredRequestIds.add(request.id);
+				this._registeredRequestIds.add(request.id);
 
-			this._register(request.response.onDidChange(() => {
-				callback();
-			}));
+				this._register(request.response.onDidChange(() => {
```

Comment:
registerModelProgressListener runs every time the set of models changes, but then any disposables using `this._register` in here will be retained for the lifetime of the service (so forever), I believe that would cause us to run these listeners multiple times per change and leak memory when chatmodels are disposed.